File: cpp/19.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <iterator>
#include <utility>
#include <map>
#include <set>
#include <string>
#include <unordered_map>
#include <climits>
#include <numeric>
using namespace std;

template<typename T>
void prints(vector<T> &v){
  cout << "[";
  for(uint i=0; i<v.size()-1; ++i)
    cout << v[i] << ", ";
  if(v.size()) cout << v[v.size()-1];
  cout << "]\n";
}

struct ListNode {
  int val;
  ListNode *next;
  ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
  ListNode *removeNthFromEnd(ListNode *head, int n) {
    ListNode *f = nullptr, *t = head;
    for(int i=0; i<n; ++i)
      t = t->next;
    while(t)
      f = f ? f->next : head, t = t->next;
    if(f){
      f->next = f->next->next;
      return head;
    } else
      return head->next;
  }
};

int main(){
  auto sol = Solution();
  ListNode *head = new ListNode(1);
  head->next = new ListNode(2);
  sol.removeNthFromEnd(head, 1);
}

File: python/19.py
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    def removeNthFromEnd(self, head, n):
        """
        :type head: ListNode
        :type n: int
        :rtype: ListNode
        """
        p = head
        ln = 0
        while p:
            ln += 1
            p = p.next

        k = ln - n
        h = ListNode(None)
        h.next = head
        p = h
        while k:
            k -= 1
            p = p.next
        p.next = p.next.next
        return h.next
